tip.lic

```diff
@@ -0,0 +1,87 @@
+=begin
+  Documentation: https://elanthipedia.play.net/Lich_script_repository#tip
+=end
+
+custom_require.call %w[common common-money]
+
+class Tip
+
+  # include DRC for the bput function
+  # include DRCM for various money handling functions
+  
+  def initialize
+    arg_definitions = [
+      # Two required arguments - recipient and amount; One optional argument, currency
+      # Mirrors DR's mechanics of "tip <person> 10000 kron", and also allows "tip <person> 10000" and assumes coppers
+      [
+        { name: 'recipient', regex: /^[A-z\.\s\-']+$/i, variable: true, description: 'The recipent of your tip. Abbreviations ok.' },
+        { name: 'amount', regex: /^\d+$/i, variable: true, description: 'Numeric amount to tip, no commas (e.g. the number "100" in 100 kronars)' },
+        { name: 'currency', regex: Regexp.union($CURRENCY_REGEX_MAP.values), optional: true, variable: true, description: 'Optional currency. Abbreviations ok: kron, lirums, dokor' }         
+      ],
+      # Three required arguments - recipient, amount, denomination - plus a fourth optional argument - currency
+      # Allows mechanic of "tip person 10 plat" and chooses your current max on hand
+      # or "tip person 10 plat kron"
+      [
+        { name: 'recipient', regex: /^[A-z\.\s\-']+$/i, variable: true, description: 'The recipent of your tip. Abbreviations ok.' },
+        { name: 'amount', regex: /^\d+$/i, variable: true, description: 'Numeric amount to tip, no commas (e.g. the number "100" in 100 kronars)' },
+        { name: 'denomination', regex: Regexp.union($DENOMINATION_REGEX_MAP.values), variable: true, description: 'The denomination. Abbrevations ok. (e.g. plat, gold, silv)' },
+        { name: 'currency', regex: Regexp.union($CURRENCY_REGEX_MAP.values), optional: true, variable: true, description: 'Optional currency. Abbreviations ok: kron, lirums, dokor' }
+      ]
+    ]
+    
+    args = parse_args(arg_definitions)  
+    gratuity(args.recipient, args.amount, args.denomination, args.currency)
+
+  end
+
+  @tip_success_patterns = [
+    /^You offer/i
+  ]
+
+  @tip_failure_patterns = [
+    /^I don't know who/i,
+    /you really should keep every bronze/i,
+    /^You already have a tip offer outstanding/i,
+    /^But you don't have that much/i,
+    /^Hmmm, tipping yourself/i    
+  ]
+
+  def gratuity(recipient, amount, denomination, currency)    
+    # Grab the total amount in copper given a numerical amount
+    # and a denomination (e.g. 100 bronze). Utilize the
+    # helper method in common-money. If no denomination given,
+    # set to copper.
+    denomination ||= 'copper'
```

Comment:
The idiomatic way is to do this in the function argument list (and avoid re-assigning to a function input), but it's pretty minor

Reply:
I agree, and that's what I did originally; however, with this method it accepts 4 non-optional arguments and it's the 3rd argument that I'm trying to provide a default for if `nil` is passed in.

For example, `gratuity('katoak', 50, nil, nil)` is what will get called so `nil` will get assigned to the arguments. Specifying the default in the function argument list would not have an effect here since the 3rd and 4th arguments are being provided.